dev/server.ts

```diff
@@ -0,0 +1,159 @@
+    const tsUrl = path.sep + relPath;
+    if (compileCache.has(tsUrl)) {
+      console.debug('Dirtying compile cache', tsUrl);
+    }
+    compileCache.delete(tsUrl);
+  }
+}
+
+/**
+ * Handler to dirty the listing cache for directory changes and .spec.ts changes.
+ * Also dirties the compile cache for changed files.
+ */
+function dirtyListingAndCompileCache(absPath: string, stats?: fs.Stats) {
+  const relPath = path.relative(srcDir, absPath);
+
+  const segments = relPath.split(path.sep);
+  // The listing changes if the directories change, or if a .spec.ts file is added/removed.
+  const listingChange =
+    (path.extname(relPath) === '' || relPath.endsWith('.spec.ts')) && segments.length > 0;
+  if (listingChange) {
+    const suite = segments[0];
+    if (listingCache.has(suite)) {
+      console.debug('Dirtying listing cache', suite);
+    }
+    listingCache.delete(suite);
+  }
+
+  dirtyCompileCache(absPath, stats);
+}
+
+watcher.on('add', dirtyListingAndCompileCache);
+watcher.on('unlink', dirtyListingAndCompileCache);
+watcher.on('addDir', dirtyListingAndCompileCache);
+watcher.on('unlinkDir', dirtyListingAndCompileCache);
+watcher.on('change', dirtyCompileCache);
+
+const app = express();
+
+// Set up logging
+app.use(morgan('dev'));
+
+// Serve the standalone runner directory
+app.use('/standalone', express.static(path.resolve(srcDir, '../standalone')));
+
+// Serve a suite's listing.js file by crawling the filesystem for all tests.
+app.get('/out/:suite/listing.js', async (req, res, next) => {
+  const suite = req.params['suite'];
+
+  if (listingCache.has(suite)) {
+    res.setHeader('Content-Type', 'application/javascript');
+    res.send(listingCache.get(suite));
+    return;
+  }
+
+  try {
+    const listing = await makeListing(path.resolve(srcDir, suite, 'listing.ts'));
+    const result = `export const listing = Promise.resolve(${JSON.stringify(
+      listing,
+      undefined,
+      2
+    )})`;
+
+    listingCache.set(suite, result);
+    res.setHeader('Content-Type', 'application/javascript');
+    res.send(result);
+  } catch (err) {
+    next(err);
+  }
+});
+
+// Serve all other .js files by fetching the source .ts file and compiling it.
+app.get('/out/**/*.js', async (req, res, next) => {
+  const tsUrl = path.relative('/out', req.url).replace(/\.js$/, '.ts');
+  if (compileCache.has(tsUrl)) {
+    res.setHeader('Content-Type', 'application/javascript');
+    res.send(compileCache.get(tsUrl));
+    return;
+  }
+
+  const absPath = path.join(srcDir, tsUrl);
+
+  try {
+    const result = await babel.transformFileAsync(absPath, babelConfig);
+    if (result && result.code) {
+      compileCache.set(tsUrl, result.code);
+
+      res.setHeader('Content-Type', 'application/javascript');
+      res.send(result.code);
+    } else {
+      throw new Error(`Failed compile ${tsUrl}.`);
+    }
+  } catch (err) {
+    next(err);
+  }
+});
+
+const host = '0.0.0.0';
```

Comment:
Should serve on 127.0.0.1 (`lo` only, rather than all interfaces), at least by default.

Reply:
but.. why not? I believe this is the default behavior of `http-server` now and it's super nice for cross-device development.

Reply:
I guess it's fine since it's part of this project so you know all you're serving up is the stuff in the project dir. Usually I default to serving locally so I don't have to worry about exposing anything (or triggering the port scraper bot on our internal network)